design-documents/storefront/catalog/product-options-and-variants.md

```diff
@@ -0,0 +1,355 @@
+That's why, with the storefront project, we have an opportunity to resolve it.
+
+Taking into account the Magento experience (good, bad, ugly) and solutions proposed on the market.
+I could define the set of requirements that I would want to see in a new options implementation:
+
+* Options are a predefined list of possible product customizations. This list belongs to the product exclusively; it is measurable, and limited only by the business requirements. So it always easy to return even a complete set of options for rendering PDP initially. 
+
+* Variants are a unique intersection of one or may option values. The variant matrix does not belong to a product as property. The variant matrix is stored managed separately from products. The primary role of variants is to distinguish the possible intersections of options from impossible.
+The variants matrix is never meant to be returned to the storefront as is. The variants matrix will be used for filtering options allowed at the storefront after one or many options selected.
+
+### Variants, Prices & Inventory
+Variants could link the options intersection with a product, a price, or an inventory record but any of these relations is not mandatory.
+So, a variant can request data from the different domains if such data assigned.
+Such a decision brings us unseen before the level of the flexibility, since there is no difference between product price and variant price, and as the consequence variant price, could be included in B2B pricing.
+
+We do not have such behaviors either Magento 1 or 2 and as a result, the whole layer of product types such as bundle and downloadable do not support B2B prices or special prices for options.
+
+The link on a product that exists in the same domain could be done through the shared ID, which as for me makes sense since I can see a possible scenario when variation that initially had only a price will be "promoted" to a product. For example due to integration reasons, to pass additional attributes data along with variant to Google Merchant Center.
+
+### Enhanced option values
+
+To reach a better user experience,
+the option could be extended with an image that represents it 
+or info URL that provides additional information about option value.
+
+[Magento: Swatches](https://docs.magento.com/user-guide/catalog/swatches.html)
+
+[Magento: Downloadable products, options with samples](https://docs.magento.com/user-guide/catalog/product-create-downloadable.html)
+
+Both types of resources could be specified as attributes of `ProductOptionValue`.
+
+```proto
+syntax = "proto3";
+
+message Product {
+    repeated ProductOption options = 100;
+}
+
+message ProductOptionValue {
+    string id = 1;
+    string label = 2;
+    string sortOrder = 3;
+    string isDefault = 4;
+    string imageUrl = 5;
+    string infoUrl = 6;
+}
+
+message ProductOption {
+    string id = 1;
+    string label = 2;
+    string sortOrder = 3;
+    string isRequired = 4;
+    string renderType = 6;
```

Comment:
Could you explain one more time how would we distinguish between custom and configurable options?

Reply:
From the storefront standpoint, no how., Because all the product variants could be represented through the one or many option values. So you do not really care how the API searches the variant, and in some way, what exactly this variant represents. The only thing about it cares is how efficiently select the option. Because  Magento does not, and should not distinguish behavior between configurable option selection and a dropdown option for instance.